# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнила студентка гр. 1304 Виноградова Маргарита. Вариант №31.

## Цель работы


Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

Аддитивной цепочкой для натурального числа $n$ называют последовательностью натуральных чисел:  
<br>
\begin{align}
    1 = a_0, a_1, a_2, … , a_r = n  
\end{align}
<br>
где каждый элемент последовательности равен сумме каких-либо двух предыдущих:   
<br>
\begin{align}
    a_i = a_j + a_k \qquad k \leqslant j < i \qquad i = 1,2,3,…,r
\end{align}
<br>

### Аддитивное дерево
* $l(n) = r$ - наименьшая длина $r$, для которой существует аддитивная цепочка для $n$
* Неравенство для метода множителей:
<br>
\begin{align}
    l(mn) \leqslant l(m) + l(n)
\end{align}
<br>
* m - равный метод $ m=2^k, n = \sum_{j = 0}^t d_j m^{t-j}$  
<br>
\begin{align}
		l(n) \leqslant m - 2 + (k + 1)t
\end{align}
<br>
* Бинарный метод $“SX”$:
<br>
\begin{align}
    l(n) \leqslant \lambda(n) + \nu(n) - 1
\end{align}
<br>
* Полагаем, что все аддитивные цепочки возрастают:
<br>
\begin{align}
    1 = a_0 < a_1 < … < a_r = n
\end{align}
<br>
* Если два числа из {$a_i$} одинаковые, то одно из них может быть опущено
* Пара $(j,k)$, $0 \leqslant k \leqslant j < i$ называется шагом $i$
* Если существует более чем одна пара $(j,k)$ полагаем, что $j$ - наибольший из возможных

#### Виды:
* Удвоение: $j = k = i - 1$
* Звездный шаг: $j = i - 1$
* Малый шаг: $\lambda(a_i) = \lambda(a_{i-1})$

#### Свойства:
* Шаг 1 всегда удвоение
* Удвоение - звездный шаг, но никогда не малый шаг 
* За удвоением всегда следует звездный шаг
* Если $i$ - ый шаг не малый, то $(i + 1)$ - ый шаг либо малый, либо звездный, либо и тот и другой
* Если $(i + 1)$ - ый шаг ни звездный, ни малый, то $i$ - ый шаг должен быть малым 


### Теорема:
Если аддитивная цепочка включает $d$ удвоений и $f = r - d$ неудвоений, то $n \leqslant 2^{d-1}F_{f + 3}$, где $F_j$ - число Фибоначи

#### Следствие:
Если аддитивная цепочка включает $f$ неудвоений и $s$ малых шагов, то
<br>
\begin{align}
    s \leqslant f \leqslant \frac{s}{1 - log_2 \phi}
\end{align}
<br>    
где $\phi = \frac{\sqrt{5} + 1}{2}$ - золотое сечение 

<br>
Алгоритм Брауэра позволяет вычислить $n$ - ую степень за  
<br>
\begin{align}
    l_B(n) = \lambda(n)+ \frac{\lambda(n)}{\lambda (\lambda(n))} + O \Bigr(\frac{\lambda(n)\lambda(\lambda (\lambda(n))))}{(\lambda (\lambda(n)))^2}\Bigl) 
\end{align}
<br> 
умножений

* Самая короткая аддитивная цепочка для числа $n$ имеет длину не более $\lambda(n)$ 
* Доказано, что почти для всех $n$ минимальная аддитивная цепочка имеет длину $l_B(n)$

<br>

Для натурального числа $n$ при заданном натуральном числе $k$ можно построить цепочку Брауэра с помощью рекуррентный формы:  

\begin{align}
B_k (n) = \left\{
\begin{array}{cl}
1, 2, 3, ..., 2^k - 1 & n < 2^k \\
B_k(q), 2q, 4q, 8q, ..., 2^kq, n & n \geq 2^k, q = \lfloor\frac{n}{2^k}\rfloor.
\end{array}
\right.
\end{align}


* Данная цепочка будет иметь длину:
<br>
\begin{align}
    l_B(n) = j(k + 1) + 2^k - 2
\end{align}
<br>
при условии, что $j_k\leqslant \lambda(n) < (j + 1)k$
* Длина будет минимизирована для больших $n$, если положить $k = \lambda(\lambda(n)) - 2\lambda(\lambda(\lambda(n)))$

### Алгоритм Брауэра

1. Задаётся некий фиксированный параметр $к$ для рассматриваемого числа $n$. Выполняется вычисление “вспомогательных чисел”:  
\begin{array}{cl}
d = 2^k, q_1 = [\frac{n}{d}],& r_1 = n  mod  d \Rightarrow n = q_1d + r_1 & (0 \leqslant r_1 < d  ) \\
q_2 = [\frac{q_1}{d}],& r_2 = q_1  mod  d \Rightarrow q_1 = q_2d + r_2  &(0 \leqslant r_2 < d  )
\end{array}
2. Данная процедура продолжается до тех пор, пока не появится такое $q_s < d$. Следовательно $q_{s-1} = q_sd+r_s$
3. Таким образом $n$ имеет вид: 
<br>
\begin{align}
    n = 2^kq_1 + r_1 = 2^k(2^kq_2 + r_2) + r_1 = ... = 2^k(2^k(...(2^kq_s + r_s) ... )+ r_2)+ r_1
\end{align}
<br>

### Алгоритм дробления вектора индексов 
1. Внешний цикл по длинам цепочек: $\underline{l}(n) \leqslant m \leqslant\overline{l}(n)$. Выбирается число $q \in \mathbb{N}$. Пусть $q = \frac{m}{2}$
2. Внутренний цикл перебора всех ${r_1, r_2, …, r_q}$ ($q!$   шагов). На каждом шаге при фиксированной части вычисляем $a_{min}$ и $a_{max}$:
    1. Если $a_m = n$ - задача решена
    2. Если $n \notin [a_{min},a_{max}]$, то переходим к следующему набору ${r_1, r_2, ..., r_q}$
    3. Если $n \in [a_{min},a_{max}]$, то организуем внутренний цикл перебора меняющейся части ${\rho_{q + 1}, \rho_{q + 2},..., \rho_m}$. Таких наборов $\frac{m!}{q!}$ штук:
        1. Если обнаружен $a_m = n$ - задача решена
        2. Если в цикле таких векторов не оказалось, то переходим к следующей (по введенной упорядочености) фиксированой части ${r_1, r_2, ..., r_q}$.
3. Если все наборы фиксированной длины исчерпаны, то увеличиваем их длину во внешнем цикле.

#### Предположения
1. Для чисел $n < 12509$ справедливо $l(n) = l^*(n)$
2. $l(2n) = l(n)$: Минимальное число $n = 191$
3. $l(2n) < l(n)$: минимальное число $n = 375494703$
4. Гипотеза Шольца - Брауэра: $l(2^n-1) \leqslant l(n) + n - 1$
    1. Гипотеза доказана для звездных цепочек: $l^*(2^n-1) \leqslant l^*(n) + n - 1$
    2. Гипотеза справедлива для всех $n < 5784689$
    3. Равенство выполняется для $1 \leqslant n \leqslant 64$

## Постановка задачи


Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы


### Алгоритм Брауэра
##### Задача: 
Реализовать алгоритм Брауэра для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра 𝑘, сопоставить длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделать выводы.


In [1]:

#алгоритм Брауэра
def brauer_algorithm(n, k):
    
    # переменная для хранения аддитивной цепочки
    additive_chain = [] 
    d = 2**k 
    
    # начальная инициализация цепочки
    for i in range(1,d):
        additive_chain.append(i)
    
    list_q = []
    # вспомогательные числа
    list_r = [] 
    q = n // d
    r = n % d
    list_r.append(r)
        
    # заполнение массива вспомогательных чисел
    while not q < d:
        r = q % d
        q = q // d
        list_r.append(r)
        
    list_r.append(q)
        
    #переменная для вычисления элемента через вспомогательное число
    tmp_need = list_r[-1]
    
    # обработка случая когда d > n (n будет в цепочке при начальной инициализации)
    if n in additive_chain:
        return additive_chain[:n]
    
    # вычисление элементов цепочки через вспомогательные числа
    while additive_chain[-1]!=n:
        
        while not q == tmp_need * d:
            q += q
            if(q not in additive_chain):
                additive_chain.append(q) 
            
        list_r.pop()
        
        tmp_need = q + list_r[-1]
        q = tmp_need
        
        # если элемент уже есть в цепочке, то он не добавляется (так как в цепочке элементы не повторяются)
        if(q not in additive_chain):
            additive_chain.append(tmp_need)
          
    
    return additive_chain


# result=brauer_algorithm(7182, 3)

# print(result)
# print(len(result))

Для вывода результатов была написана функция print_result, значение для теста выбирались случайно, параметр $k$ варьируется от 2 до 6. Для построения таблицы для анализа изменения длины приблизительной аддитивной цепочки от параметра $k$, была создана функция print_test_table, которая содержит минимальные цепочки.

In [2]:
def print_test_table():
    print("| n | k |  Brauer additive chain | length of Brauer additive chain  |  minimal additive chain | length of minimal additive chain|")
    print("|---|---|---|---|---|---|") 
    list_n = [7182, 1007, 555, 468, 127, 45, 15]
    minimal_chains = {
        7182 : [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026, 2052, 4104, 6156, 7182],
        1007 : [1, 2, 4, 8, 16, 32, 64, 65, 130, 260, 325, 650, 975, 1007],
        555 : [1, 2, 4, 8, 16, 32, 64, 72, 136, 137, 209, 418, 555],
        468 : [1, 2, 4, 8, 16, 32, 64, 68, 136, 200, 400, 468],
        127 : [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127],
        45 :  [1, 2, 4, 8, 9, 18, 36, 45],
        15 : [1, 2, 4, 5, 10, 15]
    }
    
    for n in list_n:
        for k in range(2,7):
            res = brauer_algorithm(n, k)
    
            print("|",n,"|",k,"|",res,"|",len(res),"|",minimal_chains[n],"|",len(minimal_chains[n]),"|")
        
def print_result():
    list_n = [7182, 1007, 555, 468, 127, 45, 15]
    for n in list_n:
        print("n = ",n)
        for k in range(2,7):
            res = brauer_algorithm(n, k)
            print("  k = ", k)
            print("  Additive chain: ", res)
            print("\n")
print_result()
print_test_table()


n =  7182
  k =  2
  Additive chain:  [1, 2, 3, 4, 7, 14, 28, 56, 112, 224, 448, 896, 1792, 1795, 3590, 7180, 7182]


  k =  3
  Additive chain:  [1, 2, 3, 4, 5, 6, 7, 8, 14, 28, 56, 112, 224, 448, 896, 897, 1794, 3588, 7176, 7182]


  k =  4
  Additive chain:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 28, 56, 112, 224, 448, 896, 1792, 3584, 7168, 7182]


  k =  5
  Additive chain:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 56, 112, 224, 448, 896, 1792, 3584, 7168, 7182]


  k =  6
  Additive chain:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 112, 224, 448, 896, 1792, 3584, 7168, 7182]


n =  1007
  k =  2
  Additive chain:  [1, 2, 3, 6, 12, 15, 30, 60, 62, 124, 248, 251, 502, 1004, 1007]


  k =  

| n | k |  Brauer additive chain | length of Brauer additive chain  |  minimal additive chain | length of minimal additive chain|
|---|---|---|---|---|---|
| 7182 | 2 | [1, 2, 3, 4, 7, 14, 28, 56, 112, 224, 448, 896, 1792, 1795, 3590, 7180, 7182] | 17 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026, 2052, 4104, 6156, 7182] | 16 |
| 7182 | 3 | [1, 2, 3, 4, 5, 6, 7, 8, 14, 28, 56, 112, 224, 448, 896, 897, 1794, 3588, 7176, 7182] | 20 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026, 2052, 4104, 6156, 7182] | 16 |
| 7182 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 28, 56, 112, 224, 448, 896, 1792, 3584, 7168, 7182] | 26 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026, 2052, 4104, 6156, 7182] | 16 |
| 7182 | 5 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 56, 112, 224, 448, 896, 1792, 3584, 7168, 7182] | 40 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026, 2052, 4104, 6156, 7182] | 16 |
| 7182 | 6 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 112, 224, 448, 896, 1792, 3584, 7168, 7182] | 72 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026, 2052, 4104, 6156, 7182] | 16 |
| 1007 | 2 | [1, 2, 3, 6, 12, 15, 30, 60, 62, 124, 248, 251, 502, 1004, 1007] | 15 | [1, 2, 4, 8, 16, 32, 64, 65, 130, 260, 325, 650, 975, 1007] | 14 |
| 1007 | 3 | [1, 2, 3, 4, 5, 6, 7, 8, 15, 30, 60, 120, 125, 250, 500, 1000, 1007] | 17 | [1, 2, 4, 8, 16, 32, 64, 65, 130, 260, 325, 650, 975, 1007] | 14 |
| 1007 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 24, 48, 62, 124, 248, 496, 992, 1007] | 23 | [1, 2, 4, 8, 16, 32, 64, 65, 130, 260, 325, 650, 975, 1007] | 14 |
| 1007 | 5 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 62, 124, 248, 496, 992, 1007] | 37 | [1, 2, 4, 8, 16, 32, 64, 65, 130, 260, 325, 650, 975, 1007] | 14 |
| 1007 | 6 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 120, 240, 480, 960, 1007] | 68 | [1, 2, 4, 8, 16, 32, 64, 65, 130, 260, 325, 650, 975, 1007] | 14 |
| 555 | 2 | [1, 2, 3, 4, 8, 16, 32, 34, 68, 136, 138, 276, 552, 555] | 14 | [1, 2, 4, 8, 16, 32, 64, 72, 136, 137, 209, 418, 555] | 13 |
| 555 | 3 | [1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 69, 138, 276, 552, 555] | 16 | [1, 2, 4, 8, 16, 32, 64, 72, 136, 137, 209, 418, 555] | 13 |
| 555 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 34, 68, 136, 272, 544, 555] | 23 | [1, 2, 4, 8, 16, 32, 64, 72, 136, 137, 209, 418, 555] | 13 |
| 555 | 5 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 34, 68, 136, 272, 544, 555] | 37 | [1, 2, 4, 8, 16, 32, 64, 72, 136, 137, 209, 418, 555] | 13 |
| 555 | 6 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 128, 256, 512, 555] | 68 | [1, 2, 4, 8, 16, 32, 64, 72, 136, 137, 209, 418, 555] | 13 |
| 468 | 2 | [1, 2, 3, 4, 7, 14, 28, 29, 58, 116, 117, 234, 468] | 13 | [1, 2, 4, 8, 16, 32, 64, 68, 136, 200, 400, 468] | 12 |
| 468 | 3 | [1, 2, 3, 4, 5, 6, 7, 14, 28, 56, 58, 116, 232, 464, 468] | 15 | [1, 2, 4, 8, 16, 32, 64, 68, 136, 200, 400, 468] | 12 |
| 468 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 29, 58, 116, 232, 464, 468] | 22 | [1, 2, 4, 8, 16, 32, 64, 68, 136, 200, 400, 468] | 12 |
| 468 | 5 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 56, 112, 224, 448, 468] | 36 | [1, 2, 4, 8, 16, 32, 64, 68, 136, 200, 400, 468] | 12 |
| 468 | 6 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 112, 224, 448, 468] | 67 | [1, 2, 4, 8, 16, 32, 64, 68, 136, 200, 400, 468] | 12 |
| 127 | 2 | [1, 2, 3, 4, 7, 14, 28, 31, 62, 124, 127] | 11 | [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127] | 11 |
| 127 | 3 | [1, 2, 3, 4, 5, 6, 7, 8, 15, 30, 60, 120, 127] | 13 | [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127] | 11 |
| 127 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 28, 56, 112, 127] | 19 | [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127] | 11 |
| 127 | 5 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 48, 96, 127] | 34 | [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127] | 11 |
| 127 | 6 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 127] | 65 | [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127] | 11 |
| 45 | 2 | [1, 2, 3, 4, 8, 11, 22, 44, 45] | 9 | [1, 2, 4, 8, 9, 18, 36, 45] | 8 |
| 45 | 3 | [1, 2, 3, 4, 5, 6, 7, 10, 20, 40, 45] | 11 | [1, 2, 4, 8, 9, 18, 36, 45] | 8 |
| 45 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 45] | 18 | [1, 2, 4, 8, 9, 18, 36, 45] | 8 |
| 45 | 5 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 45] | 33 | [1, 2, 4, 8, 9, 18, 36, 45] | 8 |
| 45 | 6 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45] | 45 | [1, 2, 4, 8, 9, 18, 36, 45] | 8 |
| 15 | 2 | [1, 2, 3, 6, 12, 15] | 6 | [1, 2, 4, 5, 10, 15] | 6 |
| 15 | 3 | [1, 2, 3, 4, 5, 6, 7, 8, 15] | 9 | [1, 2, 4, 5, 10, 15] | 6 |
| 15 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15] | 15 | [1, 2, 4, 5, 10, 15] | 6 |
| 15 | 5 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15] | 15 | [1, 2, 4, 5, 10, 15] | 6 |
| 15 | 6 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15] | 15 | [1, 2, 4, 5, 10, 15] | 6 |

#### Вывод
Таким образом, был реализован алгоритм Брауэра. Также следует отметить, что алгоритм находит цепочку по длине близкой к минимальной или равной ей, за малое время работы. Из данных таблицы можно увидеть, что при возрастании параметра $k$, длина цепочки увеличивается, это связано с увеличением параметра d равного $2^k$ (так как в цепочку записываю все числа от 1 до $d-1$)

### Алгоритм дробления индексов
#### Задача
Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм минимум для 5 значений 𝑛 > 1000. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы.

In [10]:
#функция для добавления единицы в вектор индексов
def change_chain(list_ind, q):
    for i in range(len(list_ind)):

        if list_ind[len(list_ind) - i - 1] == 1: 
            list_ind[len(list_ind) - i - 1] = len(list_ind) - i + q
        else:
            list_ind[len(list_ind) - i - 1] = list_ind[len(list_ind) - i - 1] - 1
            break
    return list_ind
        
#алгоритм дробления индексов        
def crushing_indexes(n):
    #определение границ m
    min_m = floor(log(n,2)) + floor(log(str(bin(n)).count('1'),2))
    max_m = floor(log(n,2)) + str(bin(n)).count('1') - 1
    
    #внешний цикл перебора m
    for m in range(min_m, max_m + 1):
        
        q = m // 2 
        #инициализация изменяемой и фиксированой части
        list_r = [i for i in range(1,q+1)]
        list_ro = [i for i in range (q+1, m+1)]
        
        #цикл перебора по list_r
        for step in range(factorial(q)+1):
        
            tmp_chain = [1]
            for ind in list_r+list_ro:
                tmp_chain.append(tmp_chain[-1] + tmp_chain[ind-1])
            
       
            a_min = tmp_chain[q] + m - q
            a_max = tmp_chain[q] * pow(2,m-q)
        
            
            if tmp_chain[-1] == n:
                return tmp_chain
            
            
            elif n < a_min or n > a_max :
                list_r = change_chain(list_r,0)
                
                
            else:
                #внутренний цикл перебора по list_ro 
                for inner_step in range(factorial(m)//factorial(q)+1):
                    
                    tmp_chain = [1]
                    for ind in list_r+list_ro:
                        tmp_chain.append(tmp_chain[-1] + tmp_chain[ind-1])
                    
                    if tmp_chain[-1] == n:
                        return tmp_chain
                    
                    list_ro = change_chain(list_ro,q)
                    
                list_r = change_chain(list_r,0)
                list_ro = [i for i in range (q+1, m+1)]


# print("START")
# print(crushing_indexes(16))
# print("END")

Для анализа алгоритма дробления индексов была написана функция print_test_crushing_indexes, которая вычисляет 5 значений n > 1000, и сравнивает полученные результаты с алгоритмов Брауэра (лучший из k от 2 до 5).


In [13]:
import time
       
def print_test_crushing_indexes():
    list_test_n = [1360,1280, 1158, 1024, 1007]
    print("|n|Crushing indexes Algorithm| length of additive chain | time crushing algorithm | Brauer additive chain | length of Brauer additive chain| time brauer algorithm|")
    print("|---|---|---|---|---|---|---|") 
    for n in list_test_n:
        start = time.time()
        crush_res = crushing_indexes(n)
        end = time.time() - start
        best_brauer = []
        best_brauer_time = 100
        for k in range(2,6):
            if len(best_brauer) > len(brauer_algorithm(n,k)) or len(best_brauer) == 0:
                start = time.time()
                best_brauer = brauer_algorithm(n,k)
                best_brauer_time = time.time() - start
        print("|",n,"|",crush_res,"|",len(crush_res),"|",end,"|",best_brauer,"|",len(best_brauer),"|",best_brauer_time,"|")
                
    
        
print_test_crushing_indexes()

|n|Crushing indexes Algorithm| length of additive chain | time crushing algorithm | Brauer additive chain | length of Brauer additive chain| time brauer algorithm|
|---|---|---|---|---|---|---|
| 1360 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 272, 544, 1088, 1360] | 13 | 0.08810591697692871 | [1, 2, 3, 4, 5, 10, 20, 21, 42, 84, 85, 170, 340, 680, 1360] | 15 | 1.3828277587890625e-05 |
| 1280 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280] | 12 | 0.0005660057067871094 | [1, 2, 3, 4, 5, 10, 20, 40, 80, 160, 320, 640, 1280] | 13 | 1.4066696166992188e-05 |
| 1158 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 384, 386, 772, 1158] | 13 | 0.024370908737182617 | [1, 2, 3, 4, 8, 16, 18, 36, 72, 144, 288, 289, 578, 1156, 1158] | 15 | 1.3113021850585938e-05 |
| 1024 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 11 | 0.0001819133758544922 | [1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 12 | 1.1920928955078125e-05 |
| 1007 | [1, 2, 4, 8, 16, 32, 64, 65, 130, 260, 325, 650, 975, 1007] | 14 | 2763.04

|n|Crushing indexes Algorithm| length of additive chain | time crushing algorithm | Brauer additive chain | length of Brauer additive chain| time brauer algorithm|
|---|---|---|---|---|---|---|
| 1360 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 272, 544, 1088, 1360] | 13 | 0.08810591697692871 | [1, 2, 3, 4, 5, 10, 20, 21, 42, 84, 85, 170, 340, 680, 1360] | 15 | 1.3828277587890625e-05 |
| 1280 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280] | 12 | 0.0005660057067871094 | [1, 2, 3, 4, 5, 10, 20, 40, 80, 160, 320, 640, 1280] | 13 | 1.4066696166992188e-05 |
| 1158 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 384, 386, 772, 1158] | 13 | 0.024370908737182617 | [1, 2, 3, 4, 8, 16, 18, 36, 72, 144, 288, 289, 578, 1156, 1158] | 15 | 1.3113021850585938e-05 |
| 1024 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 11 | 0.0001819133758544922 | [1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 12 | 1.1920928955078125e-05 |
| 1007 | [1, 2, 4, 8, 16, 32, 64, 65, 130, 260, 325, 650, 975, 1007] | 14 | 2763.0467190742493 | [1, 2, 3, 6, 12, 15, 30, 60, 62, 124, 248, 251, 502, 1004, 1007] | 15 | 1.5020370483398438e-05 |

#### Вывод
Таким образом был реализован алгоритм дробления индексов. Как видно из таблицы, данный алгоритм за счет полного перебора находит кратчайщую цепочку, но временные затраты при этом очень большие в сравнение с алгоритмом Брауэра, который за много меньшее время дает приблизительный результат.

### Гипотеза Шольца—Брауэра
#### Задача
Проверить гипотезу Шольца—Брауэра для всех натуральных $1 \leqslant 𝑛 \leqslant 12$ на алгоритме дробления вектора индексов. Результаты оформить в виде таблицы. Сделать выводы.

Так как временые затраты при больших n велики, алгоритм был оптимизирован с использование предположения 4 C (равенство $l(2^n-1) = l(n) + n - 1$). Для оптимизации сначала вычисляется $l(n) + n - 1$ и потом данная длина передается в функцию, где перебираются значения только с одним m. Также были написаны две функции, для вывода и для построения таблицы

In [17]:

def crushing_indexes_test_hypothesis(n,m):   
    if n == 1:
        return [1]
    
    q = m // 2
    
    #инициализация изменяемой и фиксированой части
    list_r = [i for i in range(1,q+1)]
    list_ro = [i for i in range (q+1, m+1)]
    
    #цикл перебора по list_r
    for step in range(factorial(q)+1):
        
        tmp_chain = [1]
        for ind in list_r + list_ro:
            tmp_chain.append(tmp_chain[-1] + tmp_chain[ind - 1])
            
        #print(tmp_chain)
        a_min = tmp_chain[q] + m - q
        a_max = tmp_chain[q] * pow(2,m-q)
        
            
        if tmp_chain[-1] == n:
            return tmp_chain
            
            
        elif n < a_min or n > a_max :
            list_r = change_chain(list_r,0)
                
                
        else:
            #внутренний цикл перебора по list_ro
            for inner_step in range(factorial(m)//factorial(q)+1):
                    
                tmp_chain = [1]
                for ind in list_r+list_ro:
                    tmp_chain.append(tmp_chain[-1] + tmp_chain[ind-1])
                    
                if tmp_chain[-1] == n:
                    return tmp_chain
                    
                list_ro = change_chain(list_ro,q)
                    
            list_r = change_chain(list_r,0)
            list_ro = [i for i in range (q+1, m+1)]
    
        

def check_Scholz_Brauer_hypothesis():
    for n in range(1,13):
        second_statement = crushing_indexes(n)
        first_statement = crushing_indexes_test_hypothesis(2**n-1,len(second_statement) + n - 2)
        
        if len(first_statement) <=  len(second_statement) + n - 1:
            print(n,": l(2^n - 1) : ",first_statement, len(first_statement))
            print(n,": l(n) + n - 1 : ",second_statement, len(second_statement) + n - 1)
            print()
        else:
            print(n,": false", first_statement, second_statement)
            break
            
def print_table_Scholz_Brauer_hypothesis():
    print("|n|$l(2^n - 1)$|chain length|$l(n) + n - 1$|chain length|statement length|")
    print("|---|---|---|---|---|---|") 
    for n in range(1,13):
        second_statement = crushing_indexes(n)
        first_statement = crushing_indexes_test_hypothesis(2**n-1,len(second_statement) + n - 2)
        
        if len(first_statement) <=  len(second_statement) + n - 1:
            print("|",n,"|",first_statement,"|",len(first_statement),"|",second_statement,"|",len(second_statement),"|",len(second_statement) + n - 1,"|")
        else:
            
            break
check_Scholz_Brauer_hypothesis()
#print_table_Scholz_Brauer_hypothesis()

1 : l(2^n - 1) :  [1] 1
1 : l(n) + n - 1 :  [1] 1

2 : l(2^n - 1) :  [1, 2, 3] 3
2 : l(n) + n - 1 :  [1, 2] 3

3 : l(2^n - 1) :  [1, 2, 4, 6, 7] 5
3 : l(n) + n - 1 :  [1, 2, 3] 5

4 : l(2^n - 1) :  [1, 2, 4, 5, 10, 15] 6
4 : l(n) + n - 1 :  [1, 2, 4] 6

5 : l(2^n - 1) :  [1, 2, 4, 8, 10, 20, 30, 31] 8
5 : l(n) + n - 1 :  [1, 2, 4, 5] 8

6 : l(2^n - 1) :  [1, 2, 4, 8, 16, 20, 21, 42, 63] 9
6 : l(n) + n - 1 :  [1, 2, 4, 6] 9

7 : l(2^n - 1) :  [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127] 11
7 : l(n) + n - 1 :  [1, 2, 4, 6, 7] 11

8 : l(2^n - 1) :  [1, 2, 4, 8, 16, 17, 34, 68, 85, 170, 255] 11
8 : l(n) + n - 1 :  [1, 2, 4, 8] 11

9 : l(2^n - 1) :  [1, 2, 4, 8, 16, 32, 64, 72, 73, 146, 292, 438, 511] 13
9 : l(n) + n - 1 :  [1, 2, 4, 8, 9] 13

10 : l(2^n - 1) :  [1, 2, 4, 8, 16, 32, 64, 68, 136, 272, 340, 341, 682, 1023] 14
10 : l(n) + n - 1 :  [1, 2, 4, 8, 10] 14

11 : l(2^n - 1) :  [1, 2, 4, 8, 16, 32, 64, 128, 256, 288, 292, 584, 585, 1170, 1755, 2047] 16
11 : l(n) + n - 1 :  [1, 2, 4, 8, 

|n|$l(2^n - 1)$|chain length|$l(n) + n - 1$|chain length|statement length|
|---|---|---|---|---|---|
| 1 | [1] | 1 | [1] | 1 | 1 |
| 2 | [1, 2, 3] | 3 | [1, 2] | 2 | 3 |
| 3 | [1, 2, 4, 6, 7] | 5 | [1, 2, 3] | 3 | 5 |
| 4 | [1, 2, 4, 5, 10, 15] | 6 | [1, 2, 4] | 3 | 6 |
| 5 | [1, 2, 4, 8, 10, 20, 30, 31] | 8 | [1, 2, 4, 5] | 4 | 8 |
| 6 | [1, 2, 4, 8, 16, 20, 21, 42, 63] | 9 | [1, 2, 4, 6] | 4 | 9 |
| 7 | [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127] | 11 | [1, 2, 4, 6, 7] | 5 | 11 |
| 8 | [1, 2, 4, 8, 16, 17, 34, 68, 85, 170, 255] | 11 | [1, 2, 4, 8] | 4 | 11 |
| 9 | [1, 2, 4, 8, 16, 32, 64, 72, 73, 146, 292, 438, 511] | 13 | [1, 2, 4, 8, 9] | 5 | 13 |
| 10 | [1, 2, 4, 8, 16, 32, 64, 68, 136, 272, 340, 341, 682, 1023] | 14 | [1, 2, 4, 8, 10] | 5 | 14 |
| 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 288, 292, 584, 585, 1170, 1755, 2047] | 16 | [1, 2, 4, 8, 10, 11] | 6 | 16 |
| 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 272, 273, 546, 1092, 1365, 2730, 4095] | 16 | [1, 2, 4, 8, 12] | 5 | 16 |

#### Вывод
Была програмно проверена гипотеза Шольца—Брауэра, для всех натуральных $1 \leqslant 𝑛 \leqslant 12$ неравенство $l(2^n - 1)\leqslant l(n) + n - 1$ верно.

## Выводы

В данной работе програмно были реализованы алгоритм построения приближенной аддитивной цепочки Брауэра и метод дробления индексов, который находит минимальную аддитивную цепочку. Первый алгоритм работает быстро, но менее точно. Второй алгоритм гарантировано найдет минимальную цепочку, но для этого потребуется много ресурсов. Таким образом, каждый алгоритм имеет свое преимущество, Брауэра в скорости работы, а алгоритм дробления индексов в точности. Также была проверена гипотеза Шольца—Брауэра для всех натуральных $n$ в границах $1 \leqslant 𝑛 \leqslant 12$ неравенство $l(2^n - 1)\leqslant l(n) + n - 1$ верно. При реализации методов использованы системы компьютерной алгебры SageMath.